This is a eMail Spam Classifers that uses Naive Bayes supervised machine learning algorithm.

In [4]:
import os
import numpy as np
from collections import Counter
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

This function builds a Dictionary of most common 3000 words from all the email content. First it adds all words and symbols in the dictionary. Then it removes all non-alpha-numeric characters and any single character alpha-numeric characters. After this is complete it shrinks the Dictionary by keeping only most common 3000 words in the dictionary. It returns the Dictionary.

In [34]:
def create_Dictionary(root_dir):                                     #define create_Dictionary to look for root_dir
  all_words = []                                                     #make all_words to be a list
  emails = [os.path.join(root_dir,f) for f in os.listdir(root_dir)]  #make emails is the list. findind root_dir, then combine into complete path with f is define in the list of root_dir
  for text in emails:                                                #create a loop for text in emails, add the conditions below
    with open(text) as m:                                            #open texts in email
      for line in m:                                                 
        words = line.split()                                         #break the line text into list of words or numbers
        all_words += words                                           #Add the single word with the single number
  dictionary = Counter(all_words)                                    #count the elements in the strings
  unused_emails = list(dictionary)                                   #create the list with words and number of frequecy

  for item in unused_emails:                                         #create the loop for unused emails to add the conditions
    if item.isalpha() == False:                                      #apply condition if the time is not aplhabic letters
      del dictionary[item]                                           #delete if the condition is true
    elif len(item) == 1:                                             #apply condition if the word length of item is 1
      del dictionary[item]                                           #delete if the condition is true
  dictionary = dictionary.most_common(3000)                          #define the dictionary where we use only 3000 common words in dictionary
  return dictionary

The codes are used to clean up the data. By breaking down the words in the emails, we can count the frequency of the words. Each word in the email is separated for the count. The result has the count for each word. Moreover, the non-alphabetical word will be removed as well. Then, when the dictionary is run, the word with the frequency that is lower than 3000 will be removed. The final result will have the list of the common words.

This function extracts feature columns and populates their values (Feature Matrix of 3000 columns and rows equal to the number of email files). The function also analyzes the File Names of each email file and decides if it's a Spam or not based on the naming convention. Based on this the function also creates the Labelled Data Column. This function is used to extract the training dataset as well as the testing dataset and returns the Feature Dataset and the Label column.

In [37]:
def extract_features(mail_dir):                                                 #define extract_features to create loops that apply on mail_dir               
  files = [os.path.join(mail_dir,fi) for fi in os.listdir(mail_dir)]            #make files is the list. findind mail_dir, then combine into complete path with fi is define in the list of mail_dir
  features_matrix = np.zeros((len(files),3000))                                 #define features_matrix with number of rows as length of files and columns to be 3000. For example, function returns ([[0,0],..]))
  train_labels = np.zeros(len(files))                                           #create the template for train_labels with the length of files represent the numbers of columns
  count = 1;                                                                    
  docID = 0;                                                                    
  for fil in files:                                                             #create loop for fil in files to add conditions
    with open(fil) as fi:                                                       #open fil when fi meets conditions
      for i, line in enumerate(fi):                                             #use enumerate to add object and count
        if i ==2:                                                               
          words = line.split()                                                  #split into separate line when words equal 2
          for word in words:                                                    #add conditions for word in words
            wordID = 0                                                          #wordID starts with 0
            for i, d in enumerate(dictionary):                            
              if d[0] == word:                                                  #create the labels of name of columns
                wordID = i                                                      
                features_matrix[docID,wordID] = words.count(word)               #count the word in words, then apply the provide the label and count for each word
      train_labels[docID] = 0;
      filepathTokens = fil.split('/')                                           
      lastToken = filepathTokens[len(filepathTokens)-1]
      if lastToken.startswith("spmsg"):                                         #define the spam message and place them in different column
        train_labels[docID] = 1;
        count = count + 1
      docID = docID + 1
  return features_matrix, train_labels              

The codes first create the templates of the table by using the length of files to be the number of rows, and number of columns with 3000. Each word will be created with the count (frequency) with it. The codes also apply the feature to separate the spam message. 

The section is the main Program that calls the above two functions and gets executed first. First it "trains" the model using model.fit function and Training Dataset. After that it scores the Test Data set by running the Trained Model with the Test Data set. At the end it prints the model performance in terms of accuracy score.

In [36]:
TRAIN_DIR = '/content/drive/MyDrive/MSBA_Colab_2020/ML_Algorithms/Naive_Bayes/train-mails'
TEST_DIR = '/content/drive/MyDrive/MSBA_Colab_2020/ML_Algorithms/Naive_Bayes/test-mails'

In [32]:
dictionary = make_Dictionary(TRAIN_DIR)

print ("reading and processing emails from TRAIN and TEST folders")
features_matrix, labels = extract_features(TRAIN_DIR)
test_features_matrix, test_labels = extract_features(TEST_DIR)

model = GaussianNB()

print ("Training Model using Gaussian Naibe Bayes algorithm .....")
model.fit(features_matrix, labels)
print ("Training completed")
print ("testing trained model to predict Test Data labels")
predicted_labels = model.predict(test_features_matrix)
print ("Completed classification of the Test Data .... now printing Accuracy Score by comparing the Predicted Labels with the Test Labels:")
print (accuracy_score(test_labels, predicted_labels))

reading and processing emails from TRAIN and TEST folders
Training Model using Gaussian Naibe Bayes algorithm .....
Training completed
testing trained model to predict Test Data labels
Completed classification of the Test Data .... now printing Accuracy Score by comparing the Predicted Labels with the Test Labels:
0.9615384615384616


From the accuracy of 96.15%, the Naive Bayes show the independent relationships between words. Based on the high freqency words, they do not have much relationship to other words. When one word appears in that email, there will not be another word that always in the same sentence; the other words will be random. The system suggests that we cannot predict the outcome of having set of words that always go together.

In [23]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


======================= END OF PROGRAM =========================